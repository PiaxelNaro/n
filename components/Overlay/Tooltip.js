// @flow

import React from 'react';
import styled from 'styled-components';
import Overlay from '.';

const ContentWrapper = styled.span`
    border-bottom: 1px dotted #aaa;
    cursor: help;
`;

const TooltipWrapper = styled.div`
    color: #999;
    padding: 8px 6px;
    font-size: 12px;
    font-weight: normal;
    line-height: 1.5;
`;

const Tooltip = ({
    content,
    children,
    showArrow = true,
}: {
    content: any,
    children: any,
    showArrow?: boolean,
}) =>
    <Overlay.Trigger
        trigger="hover"
        getOverlay={() =>
            <Overlay showArrow={showArrow}>
                <TooltipWrapper>{content}</TooltipWrapper>
            </Overlay>}
    >
        <ContentWrapper>{children}</ContentWrapper>
    </Overlay.Trigger>;

export default Tooltip;
