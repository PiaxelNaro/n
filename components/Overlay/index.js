// @flow
import React from 'react';
import classnames from 'classnames';
import styled from 'styled-components';
import Portal from 'react-portal';
import getPositionCoords from '../../modules/getPositionCoords';

type TriggerType = 'hover' | 'click';

const ARROW_OFFSET = 30;

const ArrowUp = styled.div`
    position: absolute;
    top: -2px;
    width: 20px;
    height: 10px;
    overflow: hidden;

    ::before {
        display: block;
        position: absolute;
        top: 2px;
        left: 50%;
        width: 16px;
        height: 16px;
        content: "";
        background-color: #fff;
        border: 1px solid #999;
        border-radius: 3px;
        box-shadow: 1px 1px 5px rgba(0, 0, 0, 0.3);
        transform: rotate(45deg);
        transform-origin: 0 0;
    }

    .align-left & {
        left: ${ARROW_OFFSET}px;
        transform: translate(-50%);
    }

    .align-right & {
        right: ${ARROW_OFFSET}px;
        transform: translate(50%);
    }
`;

const OverlayWrapper = styled.div.attrs({
    className: props => classnames({ 'with-arrow': props.showArrow }),
})`
    position: absolute;
    top: 0;
    padding-top: 5px;
    min-width: 200px;
    max-width: 700px;

    &.with-arrow {
        padding-top: 7px;

    }

    .align-left &.with-arrow {
        margin-left: 50%;
        transform: translateX(-${ARROW_OFFSET}px);
    }

    .align-right &.with-arrow {
        margin-right: 50%;
        transform: translateX(${ARROW_OFFSET}px);
    }
`;

const OverlayInner = styled.div`
    box-sizing: border-box;
    border: 1px solid #999;
    background-color: #fff;
    border-radius: 2px;
    box-shadow: 1px 1px 5px rgba(0, 0, 0, 0.3);

    font-size: 12px;
    text-align: left;
    color: #2F4B66;
`;

const OverlayPositionner = styled.div.attrs({
    className: props => `align-${props.alignTo}`,
})`
    position: absolute;
    z-index: 1;
    height: 0;

    &.align-left ${OverlayWrapper} {
        left: 0;
    }

    &.align-right ${OverlayWrapper} {
        right: 0;
    }
`;

const Overlay = ({ children, showArrow }: { children?: any, showArrow?: boolean }) =>
    <OverlayWrapper showArrow={showArrow}>
        {showArrow ? <ArrowUp /> : ''}
        <OverlayInner>
            {children}
        </OverlayInner>
    </OverlayWrapper>;

const OverlayControl = styled.div`
    display: inline-block;
    position: relative;
`;
OverlayControl.displayName = 'OverlayControl';

const DEFAULT_POSITION = {
    style: {
        top: 0,
        left: 0,
    },
    alignTo: 'left',
};

Overlay.Trigger = class extends React.Component {
    props: {
        children: any,
        trigger: TriggerType,
        getOverlay: () => any,
    };

    state: {
        isOpen: boolean,
    };

    toggleOverlay: Function;
    closeOverlay: Function;
    openOverlay: Function;
    closeTimeout: any;
    triggerEl: HTMLElement;

    constructor() {
        super();

        this.state = {
            isOpen: false,
        };

        this.toggleOverlay = this.toggleOverlay.bind(this);
        this.closeOverlay = this.closeOverlay.bind(this);
        this.openOverlay = this.openOverlay.bind(this);
    }

    toggleOverlay() {
        if (this.state.isOpen) {
            this.closeOverlay();
        } else {
            this.openOverlay();
        }
    }

    closeOverlay() {
        // Because the overlay is not inside the trigger, the overlay close if we move the mouseout.
        // This prevents user from clicking buttons and links that might be inside the overlay
        // itself.
        //
        // To keep the overlay open, we use a timeout so the close operation can be cancelled when
        // the Overlay mouseOver event triggers. Example: a user move their mouse from the trigger
        // to the overlay, the closeOverlay function is called (trigger mouseOut handler), then the
        // openOverlay immediately triggers (overlay mouseOver handler) and cancel the delayed
        // close action.
        //
        // 10ms in this case is an arbitrary time. It is small enough to not be noticeable, and
        // give us a few ms in case of unforeseen delay in processing the overlay mouse hover
        // action.
        this.closeTimeout = setTimeout(() => {
            this.setState({ isOpen: false });
        }, 10);
    }

    openOverlay() {
        clearTimeout(this.closeTimeout);

        this.setState({
            isOpen: true,
        });
    }

    render() {
        const { trigger, getOverlay } = this.props;

        let overlay;
        if (this.state.isOpen) {
            let position = DEFAULT_POSITION;
            if (this.triggerEl) {
                position = getPositionCoords(this.triggerEl);
            }
            overlay = (
                <Portal isOpened closeOnOutsideClick onClose={this.closeOverlay}>
                    <OverlayPositionner
                        alignTo={position.alignTo}
                        style={position.style}
                        onMouseOver={trigger === 'hover' && this.openOverlay}
                        onMouseOut={trigger === 'hover' && this.closeOverlay}
                    >
                        {getOverlay()}
                    </OverlayPositionner>
                </Portal>
            );
        }

        return (
            <OverlayControl
                innerRef={ref => (this.triggerEl = ref)}
                onClick={trigger === 'click' && this.toggleOverlay}
                onMouseOver={trigger === 'hover' && this.openOverlay}
                onMouseOut={trigger === 'hover' && this.closeOverlay}
            >
                {this.props.children}
                {overlay}
            </OverlayControl>
        );
    }
};
Overlay.Trigger.displayName = 'OverlayTrigger';
export { OverlayWrapper, OverlayPositionner };
export default Overlay;
