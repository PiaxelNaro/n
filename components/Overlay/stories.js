// @flow
import React from 'react';
import { storiesOf } from '@kadira/storybook';
import Overlay from '.';
import Tooltip from './Tooltip';
import Button from '../Button';
import FlexGrid from '../FlexGrid';

const tooltipContent = (
    <div>
        <p>
            <a href="https://www.facebook.com/" target="_blank" rel="noopener noreferrer">
                Facebook
            </a>
        </p>
        <p>
            <a href="https://www.youtube.com/" target="_blank" rel="noopener noreferrer">
                Youtube
            </a>
        </p>
    </div>
);

storiesOf('Overlay', module)
    .add('Static Overlay', () =>
        <Overlay>
            Hello
        </Overlay>
    )
    .add('Overlay trigger', () =>
        <FlexGrid>
            <FlexGrid.Item flex={1}>
                <Overlay.Trigger
                    trigger="click"
                    getOverlay={() =>
                        <Overlay>
                            Hello
                        </Overlay>}
                >
                    <Button>Open left</Button>
                </Overlay.Trigger>
                <Overlay.Trigger
                    trigger="click"
                    getOverlay={() =>
                        <Overlay showArrow>
                            Hello
                        </Overlay>}
                >
                    <Button>Open tooltip</Button>
                </Overlay.Trigger>
            </FlexGrid.Item>
            <FlexGrid.Item flex={0}>
                <Overlay.Trigger
                    trigger="click"
                    getOverlay={() =>
                        <Overlay showArrow>
                            Hello
                        </Overlay>}
                >
                    <Button>Open tooltip</Button>
                </Overlay.Trigger>
                <Overlay.Trigger
                    trigger="click"
                    getOverlay={() =>
                        <Overlay>
                            Hello
                        </Overlay>}
                >
                    <Button>Open right</Button>
                </Overlay.Trigger>
            </FlexGrid.Item>
        </FlexGrid>
    )
    .add('Tooltip', () => <Tooltip content={tooltipContent}>Ad Platforms</Tooltip>);
