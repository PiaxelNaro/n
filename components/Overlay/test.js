// @flow
import React from 'react';
import { shallow, mount } from 'enzyme';
import Overlay, { OverlayWrapper, OverlayPositionner } from '.';

jest.useFakeTimers();

describe('<Overlay>', () => {
    it('renders children', () => {
        expect(shallow(<Overlay>My Overlay</Overlay>)).toMatchSnapshot();
    });

    it('renders as tooltip', () => {
        expect(shallow(<Overlay showArrow>My Overlay</Overlay>)).toMatchSnapshot();
    });
});

describe('<OverlayWrapper>', () => {
    it('render with arrow', () => {
        const overlayWrapper = shallow(
            <OverlayWrapper showArrow className="AAOverlay">
                My Right Overlay Wrapper
            </OverlayWrapper>
        );
        expect(overlayWrapper).toMatchSnapshot();
    });
});

describe('<OverlayPositionner />', () => {
    it('Assign right className', () => {
        const positionner = shallow(<OverlayPositionner alignTo="left" />);
        expect(positionner.prop('className')).toContain('align-left');
    });
});

describe('<Overlay.Trigger>', () => {
    it('toggle overlay', () => {
        const overlay = shallow(
            <Overlay.Trigger trigger="click" getOverlay={() => <div id="overlay" />}>
                Hello
            </Overlay.Trigger>
        );
        expect(overlay).toMatchSnapshot();

        // Toggle with button
        overlay.find('OverlayControl').simulate('click');
        expect(overlay.find('#overlay').exists()).toBe(true);
        expect(overlay).toMatchSnapshot();

        overlay.find('OverlayControl').simulate('click');
        jest.runAllTimers();
        expect(overlay.find('#overlay').exists()).toBe(false);

        // Test click outside
        overlay.find('OverlayControl').simulate('click');
        expect(overlay.find('#overlay').exists()).toBe(true);

        overlay.find('Portal').prop('onClose')();
        jest.runAllTimers();
        expect(overlay.find('#overlay').exists()).toBe(false);
    });

    it("click won't trigger toggle overlay with hover type", () => {
        const overlay = shallow(
            <Overlay.Trigger trigger="hover" getOverlay={() => <div id="overlay" />}>
                Hello
            </Overlay.Trigger>
        );
        expect(overlay).toMatchSnapshot();

        // there should be no click handler
        expect(overlay.find('OverlayControl').prop('onClick')).toEqual(false);

        overlay.find('OverlayControl').simulate('mouseOver');
        expect(overlay.find('#overlay').exists()).toBe(true);

        overlay.find('OverlayControl').simulate('mouseOut');
        jest.runAllTimers();
        expect(overlay.find('#overlay').exists()).toBe(false);
    });

    it('position overlay under the trigger element', () => {
        const overlay = shallow(
            <Overlay.Trigger trigger="click" getOverlay={() => <div id="overlay" />}>
                Hello
            </Overlay.Trigger>
        );
        const instance: any = overlay.instance();
        const triggerEl: any = {
            getBoundingClientRect: () => ({
                top: 20,
                height: 30,
                width: 200,
                left: 40,
                right: 70,
            }),
        };
        instance.triggerEl = triggerEl;

        overlay.find('OverlayControl').simulate('click');
        expect(overlay.find('Overlay__OverlayPositionner').prop('style')).toEqual({
            left: 40,
            top: 50,
            width: 200,
        });
    });

    it('assigns refs', () => {
        const overlay = mount(
            <Overlay.Trigger trigger="click" getOverlay={() => <div id="overlay" />}>
                Hello
            </Overlay.Trigger>
        );
        overlay.find('OverlayControl').simulate('click');
        expect(overlay.find('Portal').exists()).toBe(true);
    });
});
