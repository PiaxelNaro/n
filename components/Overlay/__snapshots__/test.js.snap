// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Overlay.Trigger> click won't trigger toggle overlay with hover type 1`] = `
<OverlayControl
  innerRef={[Function]}
  onClick={false}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  Hello
</OverlayControl>
`;

exports[`<Overlay.Trigger> toggle overlay 1`] = `
<OverlayControl
  innerRef={[Function]}
  onClick={[Function]}
  onMouseOut={false}
  onMouseOver={false}
>
  Hello
</OverlayControl>
`;

exports[`<Overlay.Trigger> toggle overlay 2`] = `
<OverlayControl
  innerRef={[Function]}
  onClick={[Function]}
  onMouseOut={false}
  onMouseOver={false}
>
  Hello
  <Portal
    closeOnOutsideClick={true}
    isOpened={true}
    onClose={[Function]}
    onOpen={[Function]}
    onUpdate={[Function]}
  >
    <Overlay__OverlayPositionner
      alignTo="left"
      onMouseOut={false}
      onMouseOver={false}
      style={
        Object {
          "left": 0,
          "top": 0,
        }
      }
    >
      <div
        id="overlay"
      />
    </Overlay__OverlayPositionner>
  </Portal>
</OverlayControl>
`;

exports[`<Overlay> renders as tooltip 1`] = `
<Overlay__OverlayWrapper
  showArrow={true}
>
  <Overlay__ArrowUp />
  <Overlay__OverlayInner>
    My Overlay
  </Overlay__OverlayInner>
</Overlay__OverlayWrapper>
`;

exports[`<Overlay> renders children 1`] = `
<Overlay__OverlayWrapper>
  <Overlay__OverlayInner>
    My Overlay
  </Overlay__OverlayInner>
</Overlay__OverlayWrapper>
`;

exports[`<OverlayWrapper> render with arrow 1`] = `
<div
  className="AAOverlay Overlay__OverlayWrapper-emz8m8-1 with-arrow lbmeGM"
>
  My Right Overlay Wrapper
</div>
`;
